tree_assume/src/optimizations/is_valid.rs

```diff
@@ -0,0 +1,65 @@
+use crate::schema_helpers::{Constructor, ESort, Purpose};
+use strum::IntoEnumIterator;
+
+fn rule_for_ctor(ctor: Constructor) -> Option<String> {
+    let actions = ctor.filter_map_fields(|field| match field.purpose {
+        Purpose::Static(_) => None,
+        Purpose::CapturedExpr | Purpose::SubExpr | Purpose::SubListExpr => Some(format!(
+            "({sort}IsValid {var})",
+            sort = field.sort().name(),
+            var = field.var()
+        )),
+    });
+
+    if actions.is_empty() {
+        return None;
+    }
+    let actions = actions.join("\n");
+    let pat = ctor.construct(|field| field.var());
+    let sort = ctor.sort().name();
+    Some(format!(
+        "(rule (({sort}IsValid {pat})) ({actions}) :ruleset always-run)"
+    ))
+}
+
+pub(crate) fn rules() -> Vec<String> {
+    ESort::iter()
+        .map(|sort| "(relation *IsValid (*))".replace('*', sort.name()))
+        .chain(Constructor::iter().filter_map(rule_for_ctor))
+        .collect::<Vec<_>>()
+}
+#[cfg(test)]
+use crate::ast::*;
+#[cfg(test)]
+use crate::Value;
+
+#[test]
+fn test_is_valid() -> Result<(), egglog::Error> {
+    let myloop = dowhile(
+        int(1),
+        parallel!(
+            less_than(arg(), int(3)),
+            switch!(int(0); parallel!(int(4), int(5)))
+        ),
+    );
+    let not_made_valid = sub(arg(), arg()); // this expression is indeed valid, but won't be marked as so
+    let build = format!("(ExprIsValid {myloop}) {not_made_valid}");
+    let check = format!(
+        "
+(check (ExprIsValid {num0}))
+(check (ExprIsValid {arg}))
+(check (ListExprIsValid (Cons {tup45} (Nil))))
+(fail (check (ExprIsValid {not_made_valid})))
+    ",
+        num0 = int(0),
+        arg = arg(),
+        tup45 = parallel!(int(4), int(5)),
+    );
+    crate::egglog_test(
+        &build,
+        &check,
+        vec![],
```

Comment:
Pass the loop in here to check that it runs

Reply:
Updated